README.md
# pdfmerge

_Command-line utility for merging, splicing, and rotating PDF documents._

[![Build Status](https://img.shields.io/github/actions/workflow/status/metaist/pdfmerge/.github/workflows/ci.yaml?branch=main&style=for-the-badge)](https://github.com/metaist/pdfmerge/actions)
[![pdfmerge on PyPI](https://img.shields.io/pypi/v/pdfmerge.svg?color=blue&style=for-the-badge)](https://pypi.org/project/pdfmerge)
[![Supported Python versions](https://img.shields.io/pypi/pyversions/pdfmerge?style=for-the-badge)](https://pypi.org/project/pdfmerge)

[Changelog] - [Issues] - [Documentation]

[changelog]: https://github.com/metaist/pdfmerge/blob/main/CHANGELOG.md
[issues]: https://github.com/metaist/pdfmerge/issues
[documentation]: https://metaist.github.io/pdfmerge/

## Why?

I find myself merging bits of different PDFs fairly regularly and really wanted a simple CLI way to do it.

## Install

```bash
python -m pip install pdfmerge
```

## Usage

    $ pdfmerge [-h] [--version] [-o FILE|--output FILE] [-p PASSWORD|--password PASSWORD] PATH[RULE[, RULE ...]] [PATH[RULE, ...]] ...]

- `-o`, `--output` output file (default: `output.pdf`).
- `-p`, '--password` password for encrypted files (default: empty string).
- `PATH` a file, directory, or wildcard string (e.g., `file*.pdf`) of files
  to merge.
- `RULE` an optional string indicating which pages to extract and rotate.
  The syntax for each rule is:

      [START][..][END][ROTATE]

  Where `START` and `END` are positive (1-based) or negative page numbers and
  `ROTATE` is one of `>`, `V`, or `<` indicating a clockwise rotation of
  90, 180, 270 degrees, respectively.

## Command-line Example

    $ pdfmerge -o out.pdf file1.pdf file2.pdf[3,3] file2.pdf[1V,2..-1] "other*.pdf[<]" "/path/pdf[1..4>,5]"

This example illustrates several features:

- specifying an output file (must not be any of the input files)
- merging multiple files, some more than once
- splicing parts of file using indices (1-based; negatives allowed)
- including the same page multiple times
- rotating a page or page range
- merging all the PDFs in a directory

[Read more about the options](https://metaist.github.io/pdfmerge/__main__.html)

# Python Module Usage

`pdfmerge` can also be imported into python scripts.

```python
from pdfmerge import pdfmerge
pdfmerge(["pdf-1.pdf", "pdf-2.pdf[2>]"], "output.pdf")
```

## License

[MIT License](https://github.com/metaist/pypdf/blob/main/LICENSE.md)

[ci-image]: https://travis-ci.org/metaist/pdfmerge.png?branch=master
[ci-status]: http://travis-ci.org/metaist/pdfmerge
[gh-code]: https://github.com/metaist/pdfmerge/zipball/master
[gh-issues]: https://github.com/metaist/pdfmerge/issues
[gh-issues-all]: https://github.com/metaist/pdfmerge/issues/search?q=
[gh-pdfmerge]: https://github.com/metaist/pdfmerge/blob/master/pdfmerge.py
[gh-setup]: https://github.com/metaist/pdfmerge/blob/master/releases/pdfmerge-latest-setup.exe
[osi-mit]: http://opensource.org/licenses/MIT
[pypdf]: https://pypi.python.org/pypi/pyPdf


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog] and this project adheres to [Semantic Versioning].

Sections order is: `Fixed`, `Changed`, `Added`, `Deprecated`, `Removed`, `Security`.

[keep a changelog]: http://keepachangelog.com/en/1.0.0/
[semantic versioning]: http://semver.org/spec/v2.0.0.html

---

## [Unreleased]

[unreleased]: https://github.com/metaist/pdfmerge/compare/prod...main

These are changes that are on `main` that are not yet in `prod`.

---

[#28]: https://github.com/metaist/pdfmerge/issues/28
[1.0.1]: https://github.com/metaist/pdfmerge/compare/1.0.0...1.0.1

## [1.0.1] - 2023-07-24T19:59:45Z

As a consequence of fixing [#28], the output file can now be one of the input files because we build up the result in a temporary file first.

**Fixed**

- [#28]: reading and writing to a network drive

---

[#13]: https://github.com/metaist/pdfmerge/issues/13
[#15]: https://github.com/metaist/pdfmerge/pull/15
[#19]: https://github.com/metaist/pdfmerge/pull/19
[#20]: https://github.com/metaist/pdfmerge/pull/20
[#21]: https://github.com/metaist/pdfmerge/issues/21
[#23]: https://github.com/metaist/pdfmerge/issues/23
[#24]: https://github.com/metaist/pdfmerge/issues/24
[#25]: https://github.com/metaist/pdfmerge/issues/25
[1.0.0]: https://github.com/metaist/pdfmerge/compare/0.0.7...1.0.0

## [1.0.0] - 2023-06-05T12:03:55Z

**Fixed**

- [#15]: use `isinstance()` instead of `type()` (by @begert)
- [#20]: `print` on python 3 (by @lukasbindreiter)

**Changed**

- [#19]: upgraded to python 3 (by @arjndr)
- [#23]: supported python version (3.8, 3.9, 3.10, 3.11)
- [#24]: directory structure (`src` layout, using `pyproject.toml`) and other cleanup

**Added**

- [#13]: prompt for password for encrypted PDFs (if not supplied on CLI)
- [#21]: complete docs in `--help`

**Removed**

- [#25]: Windows-specific builds
- `pdfmerge.add`, `pdfmerge.merge`

---

[#12]: https://github.com/metaist/pdfmerge/pull/12
[0.0.7]: https://github.com/metaist/pdfmerge/compare/0.0.6...0.0.7

## [0.0.7] - 2015-01-01T03:28:13Z

**Fixed**

- [#12]: installing `pdfmerge` when `PyPDF2` is not installed

**Added**

- [#12]: decrypting PDFs

---

[0.0.6]: https://github.com/metaist/pdfmerge/compare/0.0.5...0.0.6

## [0.0.6] - 2014-05-02T02:11:13Z

**Removed**

- unnecessary dependencies to make sure `pip install pdfmerge` works correctly

---

[#11]: https://github.com/metaist/pdfmerge/issues/11
[0.0.5]: https://github.com/metaist/pdfmerge/commits/0.0.5

## [0.0.5] - 2014-05-02T01:19:20Z

**Changed**

- [#11]: Upgraded to `PyPDF2`


test/TestMerge.py
#!/usr/bin/python
# coding: utf-8

import os
import unittest

import pdfmerge

TEST_PDF_DIR = os.path.join('test', 'pdfs')
TEST_PDF_1 = os.path.join(TEST_PDF_DIR, 'test-1.pdf')
TEST_PDF_2 = os.path.join(TEST_PDF_DIR, 'test-2.pdf')


class TestMerge(unittest.TestCase):
    def test_pass_single(self):
        """A non-merge."""
        writer = pdfmerge.add(TEST_PDF_1)
        self.assertTrue(writer is not None)
        self.assertTrue(1, writer.getNumPages())

    def test_merge_dir(self):
        """Merge all PDFs in a directory."""
        writer = pdfmerge.add(TEST_PDF_DIR, password='test')
        self.assertTrue(writer is not None)
        self.assertTrue(4, writer.getNumPages())

    def test_rule(self):
        """Merge part of a PDF."""
        writer = pdfmerge.add(TEST_PDF_2 + '[2]')
        self.assertTrue(writer is not None)
        self.assertTrue(1, writer.getNumPages())


test/test_pdfmerge.py
"""Test pdfmerge."""

# native
from pathlib import Path

# pkg
# import pdfmerge

TEST_PDF_DIR = Path("test") / "pdfs"
TEST_PDF_1 = TEST_PDF_DIR / "test-1.pdf"
TEST_PDF_2 = TEST_PDF_DIR / "test-2.pdf"


# def test_pass_single() -> None:
#     """A non-merge."""
#     writer = pdfmerge.add(TEST_PDF_1)
#     self.assertTrue(writer is not None)
#     self.assertTrue(1, writer.getNumPages())

# def test_merge_dir(self):
#     """Merge all PDFs in a directory."""
#     writer = pdfmerge.add(TEST_PDF_DIR, password="test")
#     self.assertTrue(writer is not None)
#     self.assertTrue(4, writer.getNumPages())

# def test_rule(self):
#     """Merge part of a PDF."""
#     writer = pdfmerge.add(TEST_PDF_2 + "[2]")
#     self.assertTrue(writer is not None)
#     self.assertTrue(1, writer.getNumPages())


src/pdfmerge/__init__.py
"""Command-line utility for merging, splicing, and rotating PDF documents.

.. include:: ../../README.md
   :start-line: 4
"""

# native
from __future__ import annotations
from dataclasses import dataclass
from dataclasses import field
from getpass import getpass
from glob import glob
from pathlib import Path
from typing import Any
from typing import Generic
from typing import List
from typing import Optional
from typing import Sequence
from typing import TypeVar
from typing import Union
import re
import sys
import tempfile
import shutil

# lib
from pypdf import PdfReader
from pypdf import PdfWriter

__all__ = ("__version__", "pdfmerge")
__version__ = "1.0.1"

ERR_PATH = "ERROR: path not found: {0}"
"""Error when a path is not found."""

ERR_RULE = "ERROR: invalid rule: {0}"
"""Error when an invalid rule is encountered."""

ERR_RANGE = "ERROR: page {0} out of range [1-{1}]"
"""Error when a page is outside of the available ranges."""

ERR_BOUNDS = "ERROR: missing upper bound on range [{0}..]"
"""Error when the upper range is missing."""

# RE_MATCH_TYPE = type(re.match("", ""))
RE_HAS_RULE = re.compile(r"^(.*)\[(.*)\]$")
"""Regex that matches when an input has a rule."""

RE_RULE = re.compile(r"^(-?\d+)?(\.\.)?(-?\d+)?([>V<])?$")
"""Regex that checks that the rule is valid."""

RULE_RANGE = ".."
"""Range indicator."""

RULE_ROTATE = {None: 0, ">": 90, "V": 180, "<": 270}
"""Rotation rules."""

RULE_DEFAULT = RULE_RANGE
"""Default rule is all pages, unrotated."""


T = TypeVar("T")
"""Generic type."""


@dataclass
class Parsed(Generic[T]):
    """Generic set of parsed results."""

    done: List[T] = field(default_factory=list)
    """Items that were parsed without error."""

    errors: List[str] = field(default_factory=list)
    """Errors encountered during parsing."""

    def __add__(self, other: Parsed[Any]) -> Parsed[Any]:
        """Merge this result with another result."""
        return Parsed(self.done + other.done, self.errors + other.errors)

    def __iadd__(self, other: Parsed[Any]) -> Parsed[Any]:
        """Update this result  with data from another result."""
        self.done += other.done
        self.errors += other.errors
        return self


@dataclass
class ParsedPath:
    """Parsed path and its rules."""

    path: Path
    """Path to process."""

    rules: List[str] = field(default_factory=list)
    """Rules to apply."""


def rangify(
    rule: Union[str, re.Match[str]],
    last: int,
) -> Union[List[int], range]:
    """Convert a rule into a range.

    Args:
        rule (str, Match): pages to extract or a regex matching the rule
        last (int): maximum number of pages

    Returns:
        Union[List[int], range]: list or `range` of pages to extract.

    Examples:
        >>> list(rangify('', 3))
        [1, 2, 3]
        >>> list(rangify('1', 10))
        [1]
        >>> list(rangify('1..3', 10))
        [1, 2, 3]
        >>> list(rangify('3..1', 10))
        [3, 2, 1]
        >>> rangify('1..', 5) == rangify('..', 5)
        True
        >>> list(rangify('-3..-1', 5))
        [3, 4, 5]
        >>> list(rangify(RE_RULE.search('5..7'), 3))
        [3]
    """
    result: Union[List[int], range]
    match: Optional[re.Match[str]]

    if isinstance(rule, str):
        match = RE_RULE.search(rule)
        assert match, ERR_RULE.format(rule)
    elif isinstance(rule, re.Match):
        assert rule is not None, ERR_RULE.format()
        match = rule

    if not match:  # pragma: no cover
        return []

    left, has_range, right, _ = match.groups()
    is_range = has_range == RULE_RANGE

    beg, end = 1, last
    if not left and not right:  # [""] => [..]
        is_range = True

    if left:
        beg = int(left)
        if beg < 1:  # too low
            beg += last + 1
        elif beg > last:  # too high
            beg = last

    if right:
        end = int(right)
        if end < 1:  # too low
            end += last + 1
        elif end > last:  # too high
            end = last
    elif is_range:
        end = last

    # Generate ranges:
    if is_range and end < beg:
        result = range(beg, end - 1, -1)
    elif is_range:
        result = range(beg, end + 1)
    else:
        result = [beg]

    return result


def add_path(
    writer: PdfWriter,
    item: ParsedPath,
    password: Optional[str] = None,
) -> PdfWriter:
    """Add some PDF pages to a PDF writer.

    Args:
        writer (PdfFileWriter): writer to add to.
        item (ParsedPath): the path and rules to add.
        password (str, optional): password for encrypted files. Defaults to `None`.

    Returns:
        PdfFileWriter: the writer object
    """
    reader = PdfReader(item.path.open("rb"))
    if reader.is_encrypted:
        if password is None:
            print(f"Reading encrypted PDF <{item.path}>")
            password = getpass()
        reader.decrypt(password)

    for rule in item.rules:
        match = RE_RULE.search(rule)
        assert match, ERR_RULE.format(rule)

        rotate = match.group(4)
        for num in rangify(match, len(reader.pages)):
            writer.add_page(reader.pages[num - 1].rotate(RULE_ROTATE[rotate]))

    return writer


def parse_paths(
    inputs: Sequence[str],
    _rule: Union[str, List[str]] = RULE_DEFAULT,
) -> Parsed[ParsedPath]:
    """Split inputs into `Path` and rules.

    Args:
        inputs (Sequence[str]): inputs to parse

        _rule (str, optional): default rule to apply. Defaults to `RULE_DEFAULT`.

    Returns:
        Parsed:
            - `.done`: contains a list of successfully parsed items
                - `.path`: `Path` to the file
                - `.rules`: list of rules to apply
            - `.errors`: list of errors encountered
    """
    result: Parsed[ParsedPath] = Parsed()

    if isinstance(_rule, str):
        _rule = [_rule]

    for item in inputs:
        ok = True
        path = None
        rules = _rule

        has_rule = RE_HAS_RULE.search(item)
        if has_rule:
            item = has_rule.group(1)
            rules = re.sub(r"\s", "", has_rule.group(2)).split(",")

        for rule in rules:
            if not RE_RULE.search(rule):
                ok, err = False, ERR_RULE.format(rule)
                if err not in result.errors:
                    result.errors.append(err)
        # rules checked

        path = Path(item)
        if path.is_dir():
            paths = [str(p) for p in path.glob("*.pdf")]
            return result + parse_paths(sorted(paths), rules)
        # folder full of PDFs handled

        if "*" in item:
            return result + parse_paths(sorted(glob(item)), rules)
        # glob handled

        if not path.exists():
            ok, err = False, ERR_PATH.format(path)
            if err not in result.errors:
                result.errors.append(err)
        # path checked

        if ok:
            result.done.append(ParsedPath(path=path, rules=rules))

    return result


def pdfmerge(inputs: List[str], output: str, password: Optional[str] = None) -> None:
    """Merge PDFs into a single PDF.

    Args:
        inputs (List[str]): list of paths to merge with optional rules
        output (str): output file name
        password (str, optional): password for encrypted files. Defaults to `None`.
    """
    parsed = parse_paths(inputs)
    if parsed.errors:
        for err in parsed.errors:
            print(err, file=sys.stderr)
        return

    with tempfile.NamedTemporaryFile(delete=False) as temp:
        writer = PdfWriter(temp)
        for item in parsed.done:
            add_path(writer, item, password)
        writer.write_stream(temp)
        temp.flush()
        shutil.move(temp.name, Path(output).expanduser().resolve())


__pdoc__ = {"pdfmerge.__main__": True}


src/pdfmerge/__main__.py
#!/usr/bin/python
# coding: utf-8
"""pdfmerge - Merge, splice, and rotate PDFs.

```
Usage: pdfmerge [--help|--version] <path>... [-p PASSWORD][-o FILE]

Options:
  -h, --help                  display this message and exit
  --version                   display the version and exit

  <path>                      paths to PDFs to merge
  -p PASS, --password PASS    password to decrypt PDFs
  -o FILE, --output FILE      output file [default: output.pdf]
```

After each `<path>`, there are some optional rules you apply. These are written
in brackets after the file name. Remember to put quotes around the file name so
that your shell doesn't interpret the brackets incorrectly.

## Use index-like notation to choose page ranges

After each `<path>`, specify which parts of the file should be spliced.

```bash
  $ pdfmerge file1.pdf[1] file2.pdf[2]
  $ pdfmerge file*.pdf[7]
  $ pdfmerge file.pdf[3..1]
  $ pdfmerge "file.pdf[1..3, 7..10]"
  $ pdfmerge "file.pdf[2, 1, 3]"
  $ pdfmerge "file.pdf[1, 4..]"
```

- If indices are omitted, all pages are included.
- Use commas to separate multiple ranges.
- Whitespace is ignored (e.g., `[1 .. 2, 3]` is the same as `[1..2,3]`), but
  remember to enclose parameters with whitespace.
- **Indices start at 1** (1-based), but negative indices are ok
    (i.e. the first page is `1`, not `0`; `-1` is the last page).
- Reverse ranges are ok (e.g., `[2..1]` is the reverse of `[1..2]`).
- Open-ended ranges are ok (e.g., `[2..]` and `[..3]`).
- Ranges outside of bounds are _quietly_ forced into bounds
    (e.g., `[2..7]` for a 3-page file is treated like `[2..3]`).

## Rotate individual pages or page ranges

After each range, use a rotator (`>`, `V`, and `<`) to rotate the range by
90, 180, and 270 degrees clockwise.

```bash
  $ pdfmerge "file.pdf[1..3>]"
  $ pdfmerge "file.pdf[1<]"
  $ pdfmerge file.pdf[1..2, 4V]
  $ pdfmerge "*.pdf[>]"
```

- When using the `<` and `>` rotators, surround the entire string
  with quotes to avoid conflicts with the command-line's use of those operators.
- The rotator is case-sensitive (i.e. it's a capital `V`).
- You cannot specify more than one rotator. (Why would you want that?)
- The rotator comes right after the range (e.g., `[1>,2..3]` not `[1>..3]`).
- You can still specify a rotator if there's no range (e.g., `[<]` is the
  same as `[1..<]`).
"""
# native
from typing import cast
from typing import List
from typing import Optional
from typing import Sequence

# lib
from attrbox import parse_docopt

# pkg
from . import __version__
from . import pdfmerge


def main(argv: Optional[Sequence[str]] = None) -> None:  # pragma: no cover
    """Main entry point."""
    args = parse_docopt(__doc__, argv, version=__version__)
    pdfmerge(
        cast(List[str], args.path),
        cast(str, args.output),
        args.password,
    )


if __name__ == "__main__":
    main()


pdfmerge.py
#!/usr/bin/python
# coding: utf-8

"""Command-line PDF utility.

Utility for merging, splicing, and rotating PDF documents.
"""

from glob import glob
import argparse
import os
import re

try:
    from PyPDF2 import PdfFileWriter, PdfFileReader
except ImportError:  # probably coming from setup.py
    pass

__author__ = 'The Metaist'
__copyright__ = 'Copyright 2015, Metaist'
__email__ = 'metaist@metaist.com'
__license__ = 'MIT'
__maintainer__ = 'The Metaist'
__status__ = 'Prototype'
__version__ = '0.0.7'
__version_info__ = tuple(__version__.split('.'))

ERROR_PATH = 'ERROR: path not found: {0}'
ERROR_RULE = 'ERROR: invalid rule: {0}'
ERROR_RANGE = 'ERROR: page {0} out of range [1-{1}]'
ERROR_BOUNDS = 'ERROR: missing upper bound on range [{0}..]'

RE_MATCH_TYPE = type(re.match('', ''))
RE_HAS_RULE = re.compile(r'^(.*)\[(.*)\]$')
RE_RULE = re.compile(r'^(-?\d+)?(\.\.)?(-?\d+)?([>V<])?$')

RULE_RANGE = '..'
RULE_ROTATE = {None: 0, '>': 90, 'V': 180, '<': 270}  # rotation rules
RULE_DEFAULT = RULE_RANGE  # all pages, unrotated


def rangify(rule, range_max=None):
    """Convert a rule into a range.

    Args:
        rule (str, obj): pages to extract or a regex matching the rule
        range_max (int): maximum number of pages

    Returns:
        (list). List of pages to extract.

    Examples:
        >>> rangify('', 3)
        [1, 2, 3]
        >>> rangify('1')
        [1]
        >>> rangify('1..3')
        [1, 2, 3]
        >>> rangify('3..1')
        [3, 2, 1]
        >>> rangify('1..', 5) == rangify('..', 5) == rangify('..5')
        True
        >>> rangify('-3..-1', 5)
        [3, 4, 5]
        >>> rangify(RE_RULE.search('5..7'), 3)
        [3]
    """
    # pylint: disable=R0912
    result, match = [], None
    if isinstance(rule, str):
        match = RE_RULE.search(rule)
        assert match, ERROR_RULE.format(rule)
    elif isinstance(rule, RE_MATCH_TYPE):
        assert rule is not None, ERROR_RULE.format()
        match = rule

    beg, isrange, end, _ = match.groups()
    isrange = (isrange == RULE_RANGE)

    if not beg and not end:
        assert range_max is not None, ERROR_BOUNDS.format(beg)
        beg, isrange, end = 1, True, range_max

    beg = (beg and int(beg)) or 1
    end = (end and int(end))

    if beg:
        beg = int(beg)
        if range_max and beg < 1:
            beg += range_max + 1
        elif range_max and beg > range_max:
            beg = range_max

    if end:
        end = int(end)
        if range_max and end < 1:
            end += range_max + 1
        elif range_max and end > range_max:
            end = range_max
    elif isrange:
        assert range_max is not None, ERROR_BOUNDS.format(beg)
        end = range_max

    # Generate ranges:
    if isrange and end < beg:
        result = sorted(range(end, beg + 1), reverse=True)
    elif isrange:
        result = range(beg, end + 1)
    else:
        result.append(beg)

    return result


def add(path, password='', writer=None, rules=RULE_DEFAULT):
    """Add one or more paths to a PdfFileWriter.

    Args:
        path (str, list):       path or list of paths to merge
        password (str):         password for encrypted files
        writer (PdfFileWriter): output writer to add pdf files
        rules (str):            pages and rotation rules

    Returns:
        (PdfFileWriter). The merged PDF ready for output.
    """
    if writer is None:
        writer = PdfFileWriter()

    if isinstance(path, list):  # merge all the paths
        for subpath in path:
            writer = add(subpath, password, writer, rules)
    else:
        match = RE_HAS_RULE.search(path)
        if match:
            path, rules = match.groups()
        rules = re.sub(r'\s', '', rules)  # remove all whitespace

        if os.path.isdir(path):  # merge all pdfs in a directory
            path = os.path.join(path, '*.pdf')

        if '*' in path:  # merge multiple files
            writer = add(glob(path), password, writer, rules)
        else:  # base case; a single file
            assert os.path.isfile(path), ERROR_PATH.format(path)
            reader = PdfFileReader(open(path, 'rb'))
            if reader.isEncrypted:
                reader.decrypt(password)

            for rule in rules.split(','):
                match = RE_RULE.search(rule)
                assert match, ERROR_RULE.format(rule)
                _, _, _, rotate = match.groups()
                for page in rangify(match, reader.getNumPages()):
                    writer.addPage(
                        reader.getPage(page - 1).rotateClockwise(
                            RULE_ROTATE[rotate]
                        )
                    )
    return writer


def merge(paths, output, password=''):  # pragma: no cover
    """Merge the paths into a single PDF.

    Args:
        paths (list): list of paths to merge
        output (str): output file name
        password (str): password for encrypted files (default: '')
    """
    writer = add(paths, password)
    with open(output, 'wb') as stream:
        writer.write(stream)


def main(args=None):    # pragma: no cover
    """Main entry point."""
    parser = argparse.ArgumentParser(description=__doc__, prog='pdfmerge',
                                     fromfile_prefix_chars='@')
    parser.add_argument('--version', action='version',
                        version='%(prog)s ' + __version__)
    parser.add_argument('paths', metavar='PATH', nargs='+',
                        help='PDF files to merge')
    parser.add_argument('-o', '--output', metavar='FILE', default='output.pdf',
                        help='output file (default: "%(default)s")')
    parser.add_argument('-p', '--password', metavar='PASSWORD', default='',
                        help='password for encrypted files (default: empty)')
    opts = parser.parse_args(args)  # command-line args parsed
    merge(opts.paths, opts.output, opts.password)  # paths merged


if __name__ == '__main__':  # pragma: no cover
    main()
